lib/prawn/document.rb

```diff
@@ -683,21 +683,20 @@ def generate_margin_box
     end
 
     def apply_margin_options(options)
-      if options[:margin]
-        # Treat :margin as CSS shorthand with 1-4 values.
-        margin = Array(options[:margin])
-        positions = { 4 => [0, 1, 2, 3], 3 => [0, 1, 2, 1],
-                      2 => [0, 1, 0, 1], 1 => [0, 0, 0, 0] }[margin.length]
-
-        [:top, :right, :bottom, :left].zip(positions).each do |p, i|
-          options[:"#{p}_margin"] ||= margin[i]
-        end
-      end
+      sides = [:top, :right, :bottom, :left]
+      margin = Array(options[:margin])
 
-      [:left, :right, :top, :bottom].each do |side|
-        if margin = options[:"#{side}_margin"]
-          state.page.margins[side] = margin
-        end
+      # Treat :margin as CSS shorthand with 1-4 values.
+      positions = { 4 => [0, 1, 2, 3], 3 => [0, 1, 2, 1],
+                    2 => [0, 1, 0, 1], 1 => [0, 0, 0, 0],
+                    0 => [] }[margin.length]
+
+      sides.zip(positions).each do |side, pos|
+        state.page.margins[side] = margin[pos] if pos
+
+        # side-specific margin options take precedence over :margin
+        specific = options["#{side}_margin"]
```

Comment:
Nice catch. We do still need to guard against `pos == nil` before doing a lookup in the margins array, and default to the last page's margin. So this an option:

```ruby
sides.zip(positions).each do |side, pos|
  new_margin = options["#{side}_margin"] || (margin[pos] if pos)
  state.page.margins[side] = new_margin if new_margin
end
```

While I'm not big on `(... if pos)`, it's not unclear. Giving `positions[0]` dummy values would prevent the presence check, but IMO that *would* be less clear. Overall I do think this is better.